synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+            raise ValueError('index already added')
+
+        s_datamodel.tlib.reqDataType(syntype)
+        iid = int.from_bytes(os.urandom(8), 'little')
+        self.indices[iid] = _MetaEntry(propname=prop, syntype=syntype, datapath=(datapath,) + args)
+        self.progresses[iid] = {'nextoffset': 0, 'ngood': 0, 'nnormfail': 0}
+        self.persist()
+
+    def delIndex(self, prop):
+        '''
+        Deletes an index
+
+        Args:
+            prop (str): the (normalized) property name
+        Returns:
+            None
+        '''
+        iid = self.iidFromProp(prop)
+        if iid is None:
+            raise ValueError('Index not present')
+        del self.indices[iid]
+        self.deleting.append(iid)
+
+        # remove the progress entry in case a new index with the same propname gets added later
+        del self.progresses[iid]
+        self.persist()
+
+    def pauseIndex(self, prop):
+        '''
+        Temporarily stop indexing one or all indices.
+
+        Args:
+            prop: (Optional[str]):  the index to stop indexing, or if None, indicate to stop all indices
+        Returns:
+            None
+
+        N.B. pausing is not persistent.  Restarting the process will resume indexing.
+        '''
+        for iid, idx in self.indices.items():
+            if prop is None or prop == idx.propname:
+                self.asleep[iid] = True
+
+    def resumeIndex(self, prop):
+        '''
+        Undo a pauseIndex.
+        Args:
+            prop: (Optional[str]):  the index to start indexing, or if None, indicate to resume all indices
+        Returns:
+            None
+        '''
+        for iid, idx in self.indices.items():
+            if prop is None or prop == idx.propname:
+                self.asleep[iid] = False
+
+    def markDeleteComplete(self, iid: int) -> None:
+        self.deleting.remove(iid)
+        self.persist()
+
+# Encodes a little endian 64-bit integer
+_Int64le = struct.Struct('<Q')
+
+def _iid_en(iid):
+    return _Int64le.pack(iid)
+
+# Decodes a little endian 64-bit integer
+def _iid_un(iid):
+    return _Int64le.unpack(iid)[0]
+
+def _inWorker(callback):
+    '''
+    Gives the decorated function to the worker to run in his thread.
+
+    (Just like inpool for the worker)
+    '''
+    def wrap(self, *args, **kwargs):
+        with s_threads.RetnWait() as retn:
+            self._workq.put((retn, callback, (self, ) + args, kwargs))
+            succ, rv = retn.wait(timeout=self.MAX_WAIT_S)
+            if succ:
+                if isinstance(rv, Exception):
+                    raise rv
+                return rv
+            raise s_exc.TimeOut()
+
+    return wrap
+
+class CryoTankIndexer:
+    '''
+    Manages indexing of a single cryotank's records
+
+    This implements a lazy indexer that indexes a cryotank in a separate thread.
+
+    Cryotank entries are msgpack-encoded json-compatible dictionaries with arbitrary nesting.  An index consists of a
```

Comment:
for a documentation sake, a record in a cryotank is not guaranteed to be JSON compatible ( we can store ``bytes``) or a dictionary.

Reply:
well, it would make it hard to run datapath over it then.

Reply:
No.  Individual values being bytes would have no effect on using data paths.  Lets keep the documentation specific to msgpack and leave json out of it ( since we can *both* use integers as dict keys *and* have fields that are bytes )